File: companion-ios/SentryKey/Cloud/CloudBackupClient.swift
import Foundation

struct CloudError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

/// Talks to the SentryKey zero-knowledge backup server. Sends only the derived
/// authKey and the already-encrypted vault envelope — never the master password.
enum CloudBackupClient {
    /// Default backup server (overridable in the app). Matches the Android default.
    static let defaultServerURL = "https://sentrykey.app"

    struct BackupMeta: Identifiable {
        let id = UUID()
        let filename: String
        let timestamp: String
        let sizeBytes: Int
    }

    static func register(baseURL: String, username: String, authKey: String) async throws {
        _ = try await request(baseURL, "/api/auth/register", method: "POST",
                              body: ["username": username, "authKey": authKey])
    }

    static func login(baseURL: String, username: String, authKey: String) async throws -> String {
        let json = try await request(baseURL, "/api/auth/login", method: "POST",
                                     body: ["username": username, "authKey": authKey])
        guard let token = json["token"] as? String, !token.isEmpty else {
            throw CloudError(message: "Login did not return a session token.")
        }
        return token
    }

    static func listBackups(baseURL: String, token: String) async throws -> [BackupMeta] {
        let json = try await request(baseURL, "/api/backups", method: "GET", token: token)
        let arr = (json["backups"] as? [[String: Any]]) ?? []
        return arr.compactMap { o in
            guard let filename = o["filename"] as? String else { return nil }
            return BackupMeta(
                filename: filename,
                timestamp: (o["timestamp"] as? String) ?? filename,
                sizeBytes: (o["sizeBytes"] as? Int) ?? 0
            )
        }
    }

    static func uploadBackup(baseURL: String, token: String, envelopeJson: String) async throws -> String {
        let json = try await request(baseURL, "/api/backups/upload", method: "POST", token: token, rawBody: envelopeJson)
        return (json["filename"] as? String) ?? ""
    }

    static func downloadBackup(baseURL: String, token: String, filename: String) async throws -> String {
        try await requestRaw(baseURL, "/api/backups/file/\(filename)", method: "GET", token: token)
    }

    // MARK: - Account recovery

    struct RecoveryStart { let otpRequired: Bool; let sentTo: [String] }
    struct RecoveryMaterial { let salt: String; let blob: String; let vault: String? }

    static func setupRecovery(baseURL: String, token: String, salt: String, blob: String,
                              authKey: String, email: String, phone: String) async throws {
        _ = try await request(baseURL, "/api/recovery/setup", method: "POST", token: token,
                              body: ["salt": salt, "blob": blob, "authKey": authKey, "email": email, "phone": phone])
    }

    static func recoveryStart(baseURL: String, username: String) async throws -> RecoveryStart {
        let j = try await request(baseURL, "/api/recovery/start", method: "POST", body: ["username": username])
        return RecoveryStart(otpRequired: (j["otpRequired"] as? Bool) ?? false, sentTo: (j["sentTo"] as? [String]) ?? [])
    }

    static func recoveryFetch(baseURL: String, username: String, otp: String?) async throws -> RecoveryMaterial {
        var body: [String: Any] = ["username": username]
        if let otp = otp, !otp.isEmpty { body["otp"] = otp }
        let j = try await request(baseURL, "/api/recovery/fetch", method: "POST", body: body)
        guard let salt = j["salt"] as? String, let blob = j["blob"] as? String else {
            throw CloudError(message: "Verification failed.")
        }
        return RecoveryMaterial(salt: salt, blob: blob, vault: j["vault"] as? String)
    }

    static func recoveryReset(baseURL: String, username: String, recoveryAuthKey: String, otp: String?,
                              newAuthKey: String, newRecovery: (salt: String, blob: String, authKey: String),
                              vault: String) async throws -> String {
        var body: [String: Any] = [
            "username": username,
            "recoveryAuthKey": recoveryAuthKey,
            "newAuthKey": newAuthKey,
            "newRecovery": ["salt": newRecovery.salt, "blob": newRecovery.blob, "authKey": newRecovery.authKey],
            "vault": vault
        ]
        if let otp = otp, !otp.isEmpty { body["otp"] = otp }
        let j = try await request(baseURL, "/api/recovery/reset", method: "POST", body: body)
        guard let token = j["token"] as? String, !token.isEmpty else {
            throw CloudError(message: "Recovery did not return a session token.")
        }
        return token
    }

    // MARK: - HTTP

    private static func normalize(_ baseURL: String) -> String {
        var s = baseURL.trimmingCharacters(in: .whitespaces)
        while s.hasSuffix("/") { s.removeLast() }
        return s
    }

    private static func request(_ baseURL: String, _ path: String, method: String,
                                token: String? = nil, body: [String: Any]? = nil,
                                rawBody: String? = nil) async throws -> [String: Any] {
        let payload: String?
        if let body = body, let data = try? JSONSerialization.data(withJSONObject: body) {
            payload = String(decoding: data, as: UTF8.self)
        } else {
            payload = rawBody
        }
        let text = try await requestRaw(baseURL, path, method: method, token: token, payload: payload)
        if text.isEmpty { return [:] }
        return (try? JSONSerialization.jsonObject(with: Data(text.utf8)) as? [String: Any]) ?? [:]
    }

    private static func requestRaw(_ baseURL: String, _ path: String, method: String,
                                   token: String? = nil, payload: String? = nil) async throws -> String {
        guard let url = URL(string: normalize(baseURL) + path) else {
            throw CloudError(message: "Invalid server URL.")
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = 30
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = token { req.setValue(token, forHTTPHeaderField: "X-Session-Token") }
        // Proves this is the official app so the server's reCAPTCHA gate lets us
        // through (the phone app can't run reCAPTCHA). Set APP_API_KEY in Info.plist
        // (from a build setting) to match the server's APP_API_KEY; empty = omitted.
        if let appKey = Bundle.main.object(forInfoDictionaryKey: "APP_API_KEY") as? String, !appKey.isEmpty {
            req.setValue(appKey, forHTTPHeaderField: "X-SentryKey-App-Key")
        }
        if let payload = payload {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = Data(payload.utf8)
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: req)
        } catch {
            throw CloudError(message: "Couldn't reach the server. Check the URL and your connection.")
        }
        let text = String(decoding: data, as: UTF8.self)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(code) else {
            let serverMsg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
            throw CloudError(message: serverMsg ?? "Server error (\(code)).")
        }
        return text
    }
}

File: companion/app/src/main/java/com/example/sentrykey/CloudBackupClient.kt
package com.example.sentrykey

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/** Surfaced to the UI with a human-readable message (often straight from the server). */
class CloudException(message: String) : Exception(message)

/**
 * Talks to the SentryKey zero-knowledge backup server (see /server). Only ever
 * sends the derived authKey and the already-encrypted vault envelope — never the
 * master password or plaintext secrets.
 */
object CloudBackupClient {
    // ⬇⬇⬇ Default backup server. Overridable in the app's Cloud Backup screen
    // (e.g. http://10.0.2.2:3000 for an emulator hitting a local dev server).
    // This points at your domain — it works once the /server app is deployed
    // there over HTTPS with DNS pointed at the host.
    const val DEFAULT_SERVER_URL = "https://sentrykey.app"

    data class BackupMeta(val filename: String, val timestamp: String, val sizeBytes: Long)

    /** Registers a new account. Throws [CloudException] on any server-side rejection. */
    suspend fun register(baseUrl: String, username: String, authKey: String, inviteCode: String) =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("username", username)
                put("authKey", authKey)
                put("inviteCode", inviteCode)
            }
            request(baseUrl, "/api/auth/register", "POST", body = body)
            Unit
        }

    /** Logs in and returns the session token. */
    suspend fun login(baseUrl: String, username: String, authKey: String): String =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("username", username)
                put("authKey", authKey)
            }
            val res = request(baseUrl, "/api/auth/login", "POST", body = body)
            res.optString("token").ifEmpty { throw CloudException("Login did not return a session token.") }
        }

    /** Lists the user's stored backups, newest first. */
    suspend fun listBackups(baseUrl: String, token: String): List<BackupMeta> =
        withContext(Dispatchers.IO) {
            val res = request(baseUrl, "/api/backups", "GET", token = token)
            val arr = res.optJSONArray("backups") ?: JSONArray()
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                BackupMeta(
                    filename = o.getString("filename"),
                    timestamp = o.optString("timestamp", o.getString("filename")),
                    sizeBytes = o.optLong("sizeBytes", 0)
                )
            }
        }

    /** Uploads an encrypted envelope; returns the server-assigned filename. */
    suspend fun uploadBackup(baseUrl: String, token: String, envelopeJson: String): String =
        withContext(Dispatchers.IO) {
            val res = request(baseUrl, "/api/backups/upload", "POST", token = token, rawBody = envelopeJson)
            res.optString("filename")
        }

    /** Downloads a backup's raw encrypted envelope JSON. */
    suspend fun downloadBackup(baseUrl: String, token: String, filename: String): String =
        withContext(Dispatchers.IO) {
            requestRaw(baseUrl, "/api/backups/file/$filename", "GET", token = token)
        }

    // ---- Account recovery ----

    data class RecoveryStart(val otpRequired: Boolean, val sentTo: List<String>)
    data class RecoveryMaterial(val salt: String, val blob: String, val vault: String?)

    /** Enrolls recovery for the signed-in user (client already wrapped the encKey). */
    suspend fun setupRecovery(
        baseUrl: String, token: String, salt: String, blob: String, authKey: String,
        email: String, phone: String
    ) = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("salt", salt); put("blob", blob); put("authKey", authKey)
            put("email", email); put("phone", phone)
        }
        request(baseUrl, "/api/recovery/setup", "POST", token = token, body = body)
        Unit
    }

    /** Begins recovery; sends an OTP if a channel is configured. */
    suspend fun recoveryStart(baseUrl: String, username: String): RecoveryStart =
        withContext(Dispatchers.IO) {
            val res = request(baseUrl, "/api/recovery/start", "POST", body = JSONObject().put("username", username))
            val arr = res.optJSONArray("sentTo") ?: JSONArray()
            RecoveryStart(res.optBoolean("otpRequired", false), (0 until arr.length()).map { arr.getString(it) })
        }

    /** Fetches the wrapped recovery material + latest vault ciphertext. */
    suspend fun recoveryFetch(baseUrl: String, username: String, otp: String?): RecoveryMaterial =
        withContext(Dispatchers.IO) {
            val body = JSONObject().put("username", username)
            if (!otp.isNullOrBlank()) body.put("otp", otp)
            val res = request(baseUrl, "/api/recovery/fetch", "POST", body = body)
            RecoveryMaterial(res.getString("salt"), res.getString("blob"), res.optString("vault").ifEmpty { null })
        }

    /** Finalizes recovery (rotate login + recovery, store re-encrypted vault); returns a session token. */
    suspend fun recoveryReset(
        baseUrl: String, username: String, recoveryAuthKey: String, otp: String?,
        newAuthKey: String, newRecoverySalt: String, newRecoveryBlob: String, newRecoveryAuthKey: String,
        vault: String
    ): String = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("username", username)
            put("recoveryAuthKey", recoveryAuthKey)
            if (!otp.isNullOrBlank()) put("otp", otp)
            put("newAuthKey", newAuthKey)
            put("newRecovery", JSONObject().apply {
                put("salt", newRecoverySalt); put("blob", newRecoveryBlob); put("authKey", newRecoveryAuthKey)
            })
            put("vault", vault)
        }
        val res = request(baseUrl, "/api/recovery/reset", "POST", body = body)
        res.optString("token").ifEmpty { throw CloudException("Recovery did not return a session token.") }
    }

    // ---- HTTP plumbing ----

    private fun normalize(baseUrl: String) = baseUrl.trim().trimEnd('/')

    /** Performs a request and parses a JSON object response; throws [CloudException] on non-2xx. */
    private fun request(
        baseUrl: String,
        path: String,
        method: String,
        token: String? = null,
        body: JSONObject? = null,
        rawBody: String? = null
    ): JSONObject {
        val text = requestRaw(baseUrl, path, method, token, body?.toString() ?: rawBody)
        return if (text.isBlank()) JSONObject() else JSONObject(text)
    }

    private fun requestRaw(
        baseUrl: String,
        path: String,
        method: String,
        token: String? = null,
        payload: String? = null
    ): String {
        val url = URL(normalize(baseUrl) + path)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 30_000
            setRequestProperty("Accept", "application/json")
            if (token != null) setRequestProperty("X-Session-Token", token)
            if (payload != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        try {
            if (payload != null) {
                conn.outputStream.use { it.write(payload.toByteArray(Charsets.UTF_8)) }
            }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
            if (code !in 200..299) {
                throw CloudException(parseError(text, code))
            }
            return text
        } catch (e: CloudException) {
            throw e
        } catch (e: Exception) {
            throw CloudException("Couldn't reach the server. Check the URL and your connection.")
        } finally {
            conn.disconnect()
        }
    }

    private fun parseError(body: String, code: Int): String =
        try {
            JSONObject(body).optString("error").ifEmpty { "Server error ($code)." }
        } catch (e: Exception) {
            "Server error ($code)."
        }
}
